### PR TITLE
Add LaplaceBase class

### PR DESCRIPTION
**Pull request by andrea-borio**
This PR has the objective of creating a common base class for Laplace solvers that are based on FEM and VEM.

### CODE REVIEW

File: src/coreComponents/physicsSolvers/simplePDE/LaplaceBase.hpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+{
+public:
+  // The default nullary constructor is disabled to avoid compiler auto-generation:
+  LaplaceBase() = delete;
+
+  // The constructor needs a user-defined "name" and a parent Group (to place this instance in the tree structure of classes)
+  LaplaceBase( const string & name,
+               Group * const parent );
+
+  // Destructor
```
Comment (rrsettgast) [inline src/coreComponents/physicsSolvers/simplePDE/LaplaceBase.hpp:39]: These are all not deoxygen comments. use ///

**Review by rrsettgast — APPROVED** [review]

File: src/coreComponents/physicsSolvers/simplePDE/LaplaceBase.hpp (hunk 1 above)
Comment (andrea-borio) [reply src/coreComponents/physicsSolvers/simplePDE/LaplaceBase.hpp]: I turned them to Doxygen comments

File: src/docs/sphinx/developerGuide/KeyComponents/AddingNewSolver.rst [hunk 2]
```diff
@@ -30,38 +30,43 @@ before diving into specifying a new solver class that meets our needs.
 
 Declaration file (reference)
 ----------------------------
-The four included headers are:
+The two included headers are:
+
+ - ``physicsSolver/simplePDE/LaplaceBase.hpp`` which declares the base class ``LaplaceBase``, shared by all
+   Laplace solvers;
```
Comment (castelletto1) [inline src/docs/sphinx/developerGuide/KeyComponents/AddingNewSolver.rst:36]: See my comment about H1-conforming formulations.

File: src/docs/sphinx/developerGuide/KeyComponents/AddingNewSolver.rst [hunk 3]
```diff
@@ -79,15 +84,22 @@ to build the global file-system like structure of GEOSX (see :ref:`GroupPar` for
 It can also be noted that the nullary constructor is deleted on purpose to avoid compiler
 automatic generation and user misuse.
 
-The next method ``CatalogName()`` is static and return the key to be added to the *Catalog* for this type of solver
+The next method ``catalogName()`` is static and return the key to be added to the *Catalog* for this type of solver
```
Comment (francoishamon) [inline src/docs/sphinx/developerGuide/KeyComponents/AddingNewSolver.rst:87]:
```suggestion
The next method ``catalogName()`` is static and returns the key to be added to the *Catalog* for this type of solver
```

File: src/docs/sphinx/developerGuide/KeyComponents/AddingNewSolver.rst [hunk 4]
```diff
@@ -139,13 +157,20 @@ In the following section, we will see where this binding takes place.
...
 
+As we see, it calls the ``LaplaceBase`` constructor, that is implemented as follows.
+
+.. literalinclude:: ../../../../coreComponents/physicsSolvers/simplePDE/LaplaceBase.cpp
+   :language: c++
+   :start-after: //START_SPHINX_INCLUDE_01
+   :end-before: //END_SPHINX_INCLUDE_01
+
 Checking out the constructor, we can see that the use of a ``registerWrapper<T>(...)``
 allow us to register the key value from the `enum` ``viewKeyStruct`` defining them as:
```
Comment (francoishamon) [inline src/docs/sphinx/developerGuide/KeyComponents/AddingNewSolver.rst:175]:
```suggestion
allows us to register the key value from the `enum` ``viewKeyStruct`` defining them as:
```

**Review by francoishamon — APPROVED** [review]
Thanks @andrea-borio for the refactoring. I only added minor comments.

File: src/coreComponents/physicsSolvers/simplePDE/LaplaceBase.cpp [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+/* STEPPING IN TIME
+   Here, we decide how we march in time in the resolutions based on the possible three options set
+   in the XML file (Steady state or Implicit transient). In the case of Implicit transient, we
+   perform an implicit step (backward Euler). The implementation of the Implicit Step is found in
+   the SolverBase.  From now on, we oscillate between specific Laplace solver operations if
+   implemented and more generic SolverBase operations.  The initial values of the solver step are
+   all at time_n, and the solver attempts to advance by a time step of dt.  This dt time step size
+   is specified initially by the user; and unfortunately, it can sometimes be too large for
+   convergence.  The SolverStep method thus returns the time step value that was actually capable of
+   solving for with good convergence.
```
[the remark this answers is not in the document]
Comment (andrea-borio) [reply src/coreComponents/physicsSolvers/simplePDE/LaplaceBase.cpp:98]: fixed the text.

File: src/coreComponents/physicsSolvers/simplePDE/LaplaceBase.cpp [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+#include "mesh/mpiCommunications/CommunicationTools.hpp"
+
+namespace geosx
+{
+
+namespace dataRepository
+{
+namespace keys
+{}
+}
```
[the remark this answers is not in the document]
Comment (andrea-borio) [reply src/coreComponents/physicsSolvers/simplePDE/LaplaceBase.cpp:32]: yes you are right

File: src/coreComponents/physicsSolvers/simplePDE/LaplaceBase.cpp [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+void LaplaceBase::setupDofs( DomainPartition const & GEOSX_UNUSED_PARAM( domain ),
+                             DofManager & dofManager ) const
+{
+  dofManager.addField( m_fieldName,
+                       DofManager::Location::Node );
+
+  dofManager.addCoupling( m_fieldName,
+                          m_fieldName,
+                          DofManager::Connector::Elem );
+}
```
[the remark this answers is not in the document]
Comment (andrea-borio) [reply src/coreComponents/physicsSolvers/simplePDE/LaplaceBase.cpp:131]: You are right. We could put an intermediate class below `LaplaceBase` and have `LaplaceVEM` and `LaplaceFEM` inheriting from it. Too many layers?

**Review by castelletto1 — APPROVED** [review]

**Pull request merged**